ev3dev.go

```diff
@@ -633,3 +637,89 @@ func setAttributeOf(d Device, attr, data string) error {
 	}
 	return nil
 }
+
+var (
+	isTesting bool
+
+	// files and fileRegLock record files that have been opened
+	// during the life of the program. There is currently no
+	// mechanism to remove a file from the registry, but this is
+	// probably not a problem given that attached devices are
+	// extremely likely to remain attached for the life of the
+	// program.
+	fileRegLock sync.Mutex
+	files       = make(map[string]*os.File)
+)
+
+func readFile(path string) ([]byte, error) {
+	if isTesting {
+		// FIXME(kortschak): Make this work always.
+		//
+		// This horror is here to work around flakey
+		// kernel hangs that happen during testing if
+		// we use the fast path code below.
+		// The flakes appear to be in bazil.org/fuse
+		// or in FUSE itself since the behaviour is
+		// dependent on bazil.org/fuse version. The
+		// behaviour is very variable, depending on
+		// timing and debugging output.
+		//
+		// The upshot of this is that the code below
+		// is only exercised on actual devices. This
+		// is not terrible, since bugs should show up
+		// quickly and the remainder of the code is
+		// properly tested using the slow path.
+		return ioutil.ReadFile(path)
+	}
+
+	f, err := fileFor(path)
+	if err != nil {
+		return nil, err
+	}
+	if f == nil {
+		// Don't try fast path for files that already
+		// failed to read into short buffer.
+		return ioutil.ReadFile(path)
+	}
+	var buf [256]byte
+	n, err := f.ReadAt(buf[:], 0)
+	if err == nil {
+		// EV3 sysfs files are maximally 4096 byte
+		// (memory page size), but files are likely
+		// to be significantly smaller. The size of
+		// 128 bytes was suggested in ev3go/ev3dev#93,
+		// but this fails with the LED trigger files.
+		// We log if there is no error since ReadAt
+		// will always return an error if n is less
+		// than len(buf). So we catch all the cases
+		// where the file is longer, with a small number
+		// of false positives where the file is exactly
+		// the length of the buffer. Bump the length
+		// of the buffer when that happens.
+		log.Printf("ev3dev: buffer too short for %s: falling back to ioutil.ReadFile", path)
```

Comment:
Couldn't this be an error other than buffer size as well?

Reply:
No, it's a non-error. The only way you can have a nil error after a call to the `*os.File` `ReadAt` method is if the buffer is the same length as or shorter than the data being read. All the non-nil error states are handled below.